geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/MessageDispatcherTest.java

```diff
@@ -173,6 +174,32 @@ public void oldClientWillNotGetClientReAuthenticateMessage() throws Exception {
     verify(dispatcher, never()).dispatchResidualMessages();
   }
 
+
+  @Test
+  public void oldClientWillContinueToDeliverMessageIfNotified() throws Exception {
+    doReturn(false, false, true).when(dispatcher).isStopped();
+    // make sure wait time is short
+    doReturn(10000L).when(dispatcher).getSystemProperty(eq(RE_AUTHENTICATE_WAIT_TIME), anyLong());
+    doThrow(AuthenticationExpiredException.class).when(dispatcher).dispatchMessage(any());
+    when(messageQueue.peek()).thenReturn(message);
+    when(proxy.getVersion()).thenReturn(KnownVersion.GEODE_1_14_0);
+
+    Thread dispatcherThread = new Thread(() -> dispatcher.runDispatcher());
+    Thread notifyThread = new Thread(() -> dispatcher.notifyReAuthentication());
+
+    dispatcherThread.start();
+    await().until(() -> dispatcher.isWaitingForReAuthentication());
+    notifyThread.start();
+
+    dispatcherThread.join();
+    notifyThread.join();
```

Comment:
Take a look at the `ExecutorServiceRule`. It can simplify a lot of this sort of testing and will kill and dump hung threads at the end of the test to make debugging easier if there is a failure of some sort.

Reply:
Oops, rules don't work under JUnit 5. We'll need to see if it can be ported at some point.

Reply:
A well-behaved rule can work in JUnit 5. Here's how:

Add `testImplementation('org.junit.jupiter:junit-jupiter-migrationsupport')` to the `build.gradle` file.

Add `@EnableRuleMigrationSupport` to the test class.

This works as long as the rule doesn't need to be on the call stack. `ExecutorServiceRule` probably doesn't need to be on the call stack.

Reply:
Yeah, if it's junit4, I would use that. Not sure how to use that in Junit5.